### PR TITLE
kubectl whoami cmd

### LINKED ISSUE #65798: Need a "whoami" sub-command for kubectl
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug

/kind feature

**What happened**:

We might want to know informations about client's identity via "kubectl". Especially the name and the groups in `UserInfo`.😃

**What you expected to happen**:

```
$ kubectl whoami
USER           GROUPS        blah..
foo               bar,baz
```

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by yue9944882**
**What this PR does / why we need it**:

Per issue.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #65798

**Special notes for your reviewer**:

**Release note**:

```release-note
Implement "whoami" command for kubectl
```

### CODE REVIEW

**Comment by yue9944882** [comment]
/sig cli
@kubernetes/sig-cli-maintainers

**Comment by liggitt** [comment]
@kubernetes/sig-auth-api-reviews

**Comment by liggitt** [comment]
xref https://github.com/kubernetes/kubernetes/issues/30784

File: pkg/kubectl/cmd/whoami.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
Comment (zjj2wry) [inline pkg/kubectl/cmd/whoami.go]: 2018

File: staging/src/k8s.io/apiserver/pkg/server/routes/whoami.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
Comment (zjj2wry) [inline staging/src/k8s.io/apiserver/pkg/server/routes/whoami.go]: ditto

File: staging/src/k8s.io/apiserver/pkg/server/routes/whoami.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/emicklei/go-restful"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+// UserIdentity provides a webservice with user identity information.
+type UserIdentity struct{}
+
+// Install registers the APIServer's `/version` handler.
```
Comment (zjj2wry) [inline staging/src/k8s.io/apiserver/pkg/server/routes/whoami.go]: version

File: staging/src/k8s.io/apiserver/pkg/server/routes/whoami.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+func (u UserIdentity) Install(c *restful.Container) {
+	versionWS := new(restful.WebService)
+	versionWS.Path("/whoami")
+	versionWS.Doc("get the information about requesting client")
+	versionWS.Route(
+		versionWS.GET("/").To(u.handleWhoAmI).
+			Doc("get the user identity information").
+			Operation("getUserIdentity").
+			Produces(restful.MIME_JSON).
+			Consumes(restful.MIME_JSON))
```
Comment (zjj2wry) [inline staging/src/k8s.io/apiserver/pkg/server/routes/whoami.go]: should be any type?

File: pkg/kubectl/cmd/whoami.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	}
+}
+
+func NewCmdWhoAmI(f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *cobra.Command {
+	o := NewWhoAmIOptions(ioStreams)
+	cmd := &cobra.Command{
+		Use:     "whoami",
+		Short:   i18n.T("Print the user's identity information"),
+		Long:    "Print the requesting user's identity for the current context",
+		Example: versionExample,
```
Comment (zjj2wry) [inline pkg/kubectl/cmd/whoami.go]: whoamiExample

File: pkg/kubectl/cmd/whoami_test.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
Comment (zjj2wry) [inline pkg/kubectl/cmd/whoami_test.go]: ditto

**Review by zjj2wry — COMMENTED** [review]
small nits, i think this command will helpful

File: pkg/kubectl/cmd/whoami.go (hunk 5 above)
Comment (yue9944882) [reply pkg/kubectl/cmd/whoami.go]: Thanks for reviewing! All done.😃

**Comment by enj** [comment]
Is there a reason to go with this approach vs. the standard REST storage APIs?  I know we probably do not need all the flexibility we get with the standard rest code, but I do not know of a reason to make this not use that "normal" path.

**Comment by yue9944882** [comment]
@enj I'm not really sure either. But `userInfo` exists in the request scope and could always changes without those changes applying to the storage. e.g. when I sign a new x509 client key-pair or use a new bearer header. So it doesn't seem to have any direct relation with the REST storage AFAIK. I'd slightly lean on make it normal path.😃

**Comment by enj** [comment]
Using the rest storage approach will make this a standard API with generated code in client-go.  Note that both approaches would be virtual so they would use the data from the request context.  The REST storage is the normal path IMO since those APIs have a very specific contract.

**Comment by liggitt** [comment]
this change needs to start with a [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/0000-kep-template.md), in order to reach agreement on the goals and approach, prior to beginning implementation

/close

**Pull request closed without merging**